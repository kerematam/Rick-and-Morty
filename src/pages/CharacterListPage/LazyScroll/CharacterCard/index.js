import React, { useCallback } from 'react'
import { useDispatch } from 'react-redux'
import PropTypes from 'prop-types'
import {
  Card,
  CardActionArea,
  CardContent,
  CardMedia,
  Typography,
} from '@material-ui/core'
import history from 'utils/history'
import {
  actions,
  useCharacterDetailReducerInjector,
} from 'pages/CharacterPage/redux'

import styles from './CharacterCard.module.scss'

export default function CharacterCard({ character }) {
  useCharacterDetailReducerInjector()
  const dispatch = useDispatch()
  const handleClick = useCallback(() => {
    dispatch(actions.setActiveCharacter({ character }))
    history.push(`/characters/${character.id}`)
  }, [dispatch, character])

  return (
    <Card square className={styles.root}>
      <CardActionArea onClick={handleClick}>
        <CardMedia
          className={styles.media}
          image={character.image}
          title={character.name}
        />
        <CardContent className={styles.content}>
          <Typography gutterBottom variant="h5" component="h2">
            {character.name}
          </Typography>
          <Typography variant="body2" color="textSecondary" component="p">
            {character.type}
          </Typography>
        </CardContent>
      </CardActionArea>
    </Card>
  )
}

CharacterCard.propTypes = {
  character: PropTypes.object,
}
