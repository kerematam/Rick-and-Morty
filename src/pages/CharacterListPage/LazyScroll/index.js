import React from 'react'
import Grid from '@material-ui/core/Grid'
import InfiniteScroll from 'react-infinite-scroller'
import Card from './CharacterCard'

const LazyScroll = ({ characters = [], loadMore, loading, hasMore }) => {
  return (
    <InfiniteScroll pageStart={0} loadMore={loadMore} hasMore={true}>
      <Grid container spacing={3}>
        {characters.map(character => (
          <Grid key={character.id} item xs={12} sm={6} md={4} lg={3}>
            <Card character={character} />
          </Grid>
        ))}
      </Grid>
    </InfiniteScroll>
  )
}

export default React.memo(LazyScroll)
