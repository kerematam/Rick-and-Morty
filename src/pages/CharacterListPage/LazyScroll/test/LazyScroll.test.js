import React from 'react'
import renderer from 'react-test-renderer'
import LazyScroll from '../index'

const setup = setupProps => {
  const wrapper = renderer.create(<LazyScroll {...setupProps} />).toJSON()

  return {
    wrapper,
  }
}

describe('LazyScroll', () => {
  test('should render correctly', () => {
    const setupProps = {
      characters: [],
      loadMore: () => {},
      loading: false,
      hasMore: true,
    }
    const { wrapper } = setup(setupProps)
    expect(wrapper).toMatchSnapshot()
  })
})
