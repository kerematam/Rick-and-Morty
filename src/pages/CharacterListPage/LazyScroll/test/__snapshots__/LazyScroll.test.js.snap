// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyScroll should render correctly 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
  />
</div>
`;
